### PR TITLE
Completing the Lock interface

### PR DESCRIPTION
**Pull request by nfurfaro**
# Description
I recently added the lock interface to our project, but after careful review with the help of a script to auto-generate the interface, I realized it was incomplete.
This PR adds the remaining functions/events to the interface.
The next step will be to modify this to work with v4 (in its own PR).

# Issues

Fixes #
Refs #

# Checklist:

- [x] 1 PR, 1 purpose: my Pull Request applies to a single purpose
  - [ ] This PR only contains configuration changes (package.json, etc.)
  - [x] This PR only contains code changes (if configuration changes are required, do a separate PR first, then re-base)
- [x] My code follows the style guidelines of this project, including naming conventions
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] If my code adds or changes components, I have written corresponding stories with Storybook
- [x] I have performed a self-review of my own code
- [ ] If my code involves visual changes, I am adding applicable screenshots to this thread

### CODE REVIEW

File: smart-contracts/contracts/interfaces/IPublicLock.sol [hunk 1]
```diff
@@ -288,19 +288,6 @@ interface IPublicLock {
...
 
-  /**
-   * This approves _approved to get ownership of _tokenId.
-   * Note: that since this is used for both purchase and transfer approvals
-   * the approved token may not exist.
-   * @param _approved The address to approve
-   * @param _tokenId The tokenId to approve
-   * @dev Throws if lock is disabled. Throws if called by other than key owner of =r approved address.
-   */
-  function approve(
```
Comment (NickCuso) [inline smart-contracts/contracts/interfaces/IPublicLock.sol]: Why remove `approve`?

**Comment by nfurfaro** [comment]
One thing I wonder about...
The interface-generator script I used outputs lots of things we don't need. For example:
`function name() constant returns (string );` In this case, `name` is a public state variable, so a getter function is created for us automatically. However, state variables don't belong in interfaces.
Along similar lines, the function `function approve(address _approved,uint256 _tokenId) ;` is already specified in `IERC721.sol`, so I don't feel like we should also be adding it to the lock interface. Would like to get a second opinion on this. cc @NickCuso 
For reference, I will post the output of the generator script below.

**Comment by nfurfaro** [comment]
> lock@1.0.0 generate /Users/overlord/DAPPS/unlock/smart-contracts
> ../scripts/generator.js "PublicLock.sol" "Lock-ABI.json"


> lock@1.0.0 generate /Users/overlord/DAPPS/unlock/smart-contracts
> ../scripts/generator.js "PublicLock.sol" "Lock-ABI.json"

contract PublicLock.sol{function BASIS_POINTS_DEN() constant returns (uint256 );
function address2Str(address _addr) constant returns (string );
  function approve(address _approved,uint256 _tokenId) ;
function balanceOf(address _owner) constant returns (uint256 );
function beneficiary() constant returns (address );
function cancelAndRefund() ;
function cancelAndRefundFor(address _keyOwner,bytes _signature) ;
function destroyLock() ;
function disableLock() ;
function erc1820() constant returns (address );
function expirationDuration() constant returns (uint256 );
function expireKeyFor(address _owner) ;
function freeTrialLength() constant returns (uint256 );
function fullRefund(address _keyOwner,uint256 amount) ;
function getApproved(uint256 _tokenId) constant returns (address );
function getBalance(address _tokenAddress,address _account) constant returns (uint256 );
function getCancelAndRefundApprovalHash(address _keyOwner,address _txSender) constant returns (bytes32 approvalHash);
function getCancelAndRefundValueFor(address _owner) constant returns (uint256 refund);
function getHasValidKey(address _owner) constant returns (bool );
function getOwnersByPage(uint256 _page,uint256 _pageSize) constant returns (address[] );
function getTokenIdFor(address _account) constant returns (uint256 );
function getTransferFee(address _owner) constant returns (uint256 );
function grantKeys(address[] _recipients,uint256[] _expirationTimestamps) ;
function invalidateApprovalToCancelKey() ;
function isAlive() constant returns (bool );
function isApprovedForAll(address _owner,address _operator) constant returns (bool );
function isKeyOwner(uint256 _tokenId,address _owner) constant returns (bool );
function isOwner() constant returns (bool );
function keyCancelInterfaceId() constant returns (bytes32 );
function keyExpirationTimestampFor(address _owner) constant returns (uint256 timestamp);
function keyOwnerToNonce(address ) constant returns (uint256 );
function keyPrice() constant returns (uint256 );
function keySoldInterfaceId() constant returns (bytes32 );
function maxNumberOfKeys() constant returns (uint256 );
function name() constant returns (string );
function numberOfOwners() constant returns (uint256 );
function owner() constant returns (address );
function ownerOf(uint256 ) constant returns (address );
function owners(uint256 ) constant returns (address );
function purchase(uint256 _value,address _recipient,address _referrer,bytes _data) ;
function refundPenaltyBasisPoints() constant returns (uint256 );
function renounceOwnership() ;
function safeTransferFrom(address _from,address _to,uint256 _tokenId) ;
function safeTransferFrom(address _from,address _to,uint256 _tokenId,bytes _data) ;
function setApprovalForAll(address _to,bool _approved) ;
function setBaseTokenURI(string _baseTokenURI) ;
function strConcat(string _a,string _b,string _c,string _d) constant returns (string _concatenatedString);
function supportsInterface(bytes4 interfaceId) constant returns (bool );
function symbol() constant returns (string );
function tokenAddress() constant returns (address );
function tokenByIndex(uint256 _index) constant returns (uint256 );
function tokenOfOwnerByIndex(address _owner,uint256 _index) constant returns (uint256 );
function tokenURI(uint256 _tokenId) constant returns (string );
function totalSupply() constant returns (uint256 );
function transferFeeBasisPoints() constant returns (uint256 );
function transferFrom(address _from,address _recipient,uint256 _tokenId) ;
function transferOwnership(address newOwner) ;
function uint2Str(uint256 _i) constant returns (string _uintAsString);
function unlockProtocol() constant returns (address );
function updateBeneficiary(address _beneficiary) ;
function updateKeyPrice(uint256 _keyPrice) ;
function updateLockName(string _lockName) ;
function updateLockSymbol(string _lockSymbol) ;
function updateRefundPenalty(uint256 _freeTrialLength,uint256 _refundPenaltyBasisPoints) ;
function updateTransferFee(uint256 _transferFeeBasisPoints) ;
function withdraw(address _tokenAddress,uint256 _amount) ;
function initialize(address _owner,uint256 _expirationDuration,address _tokenAddress,uint256 _keyPrice,uint256 _maxNumberOfKeys,string _lockName) ;
function initialize() ;
function initialize(address _tokenAddress) ;
function publicLockVersion() constant returns (uint16 );
}

File: smart-contracts/contracts/interfaces/IPublicLock.sol (hunk 1 above)
Comment (nfurfaro) [reply smart-contracts/contracts/interfaces/IPublicLock.sol]: See my comment below, bumy thinking is that Approve is already defined in IERC721.sol, so we shouldn't need it in our lock interface (as PublicLock.Sol inherits both IERC721.sol and IPublicLock.Sol.)

**Comment by NickCuso** [comment]
RE `name`- adding this to the interface is adding the auto-generated getter to the interface, not adding a state variable.  So that's fine to add.

RE `approve`- I don't think it's clear for consumers atm.  Easiest would be to include everything for convenience.  Alternatively we could add comments to clarify which set of interfaces define the full contract (and make each of them available in our published package).

**Comment by nfurfaro** [comment]
Ok. I think I'll proceed with adding everything for now,( including
autogenerated getters for the state variables) and add comments noting when
a function or event originated from from a separate inherited interface.

**Comment by NickCuso** [comment]
Hmm it appears something went wrong with the history for this one.  Could you rebase?  it shouldn't be showing so much under the diff.

**Comment by nfurfaro** [comment]
@NickCuso hey, sorry for the noise. I was just trying to fix something else in my local repo and it made a mess. I'll resubmit a clean version of this with just the one file change once its done.

**Comment by nfurfaro** [comment]
Closing in order to clean up some mess. Will reopen a new PR for this.

**Pull request closed without merging**